src/Workspaces/Core/Portable/Utilities/NullableHelpers/NamedTypeSymbolWithNullableAnnotation.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal static partial class NullableExtensions
+    {
+        private sealed class NamedTypeSymbolWithNullableAnnotation : TypeSymbolWithNullableAnnotation, INamedTypeSymbol
+        {
+            public NamedTypeSymbolWithNullableAnnotation(INamedTypeSymbol wrappedSymbol, NullableAnnotation nullability) : base(wrappedSymbol, nullability)
+            {
+            }
+
+            public override void Accept(SymbolVisitor visitor)
+            {
+                visitor.VisitNamedType(this);
+            }
+
+            public override TResult Accept<TResult>(SymbolVisitor<TResult> visitor)
+            {
+                return visitor.VisitNamedType(this);
+            }
+
+            #region INamedTypeSymbol Implementation Forwards
+
+            public int Arity => WrappedSymbol.Arity;
+            public bool IsGenericType => WrappedSymbol.IsGenericType;
+            public bool IsUnboundGenericType => WrappedSymbol.IsUnboundGenericType;
+            public bool IsScriptClass => WrappedSymbol.IsScriptClass;
+            public bool IsImplicitClass => WrappedSymbol.IsImplicitClass;
+            public bool IsComImport => WrappedSymbol.IsComImport;
+            public IEnumerable<string> MemberNames => WrappedSymbol.MemberNames;
+            public ImmutableArray<ITypeParameterSymbol> TypeParameters => WrappedSymbol.TypeParameters;
+            public ImmutableArray<ITypeSymbol> TypeArguments => WrappedSymbol.TypeArguments;
+            public ImmutableArray<NullableAnnotation> TypeArgumentsNullableAnnotations => WrappedSymbol.TypeArgumentsNullableAnnotations;
+            public IMethodSymbol DelegateInvokeMethod => WrappedSymbol.DelegateInvokeMethod;
+            public INamedTypeSymbol EnumUnderlyingType => WrappedSymbol.EnumUnderlyingType;
+            public INamedTypeSymbol ConstructedFrom => WrappedSymbol.ConstructedFrom;
+            public ImmutableArray<IMethodSymbol> InstanceConstructors => WrappedSymbol.InstanceConstructors;
+            public ImmutableArray<IMethodSymbol> StaticConstructors => WrappedSymbol.StaticConstructors;
+            public ImmutableArray<IMethodSymbol> Constructors => WrappedSymbol.Constructors;
+            public ISymbol AssociatedSymbol => WrappedSymbol.AssociatedSymbol;
+            public bool MightContainExtensionMethods => WrappedSymbol.MightContainExtensionMethods;
+            public INamedTypeSymbol TupleUnderlyingType => WrappedSymbol.TupleUnderlyingType;
+            public ImmutableArray<IFieldSymbol> TupleElements => WrappedSymbol.TupleElements;
+            public bool IsSerializable => WrappedSymbol.IsSerializable;
+            private new INamedTypeSymbol WrappedSymbol => (INamedTypeSymbol)base.WrappedSymbol;
+            INamedTypeSymbol INamedTypeSymbol.OriginalDefinition => WrappedSymbol.OriginalDefinition;
```

Comment:
i'm not certain what expected behavior is for OrigDef.  If i have `IList<T>?` is the origdef `IList<T>` or do i have the origdef?

Reply:
My current operating belief is OriginalDefinition should be `IList<T>`, thinking of the nullable annotation being something not unlike a constructed generic type. I think this works out reasonably because one common use for OriginalDefinition is to then compare via reference equality to the return of Compilation.GetTypeByMetadataName, and then that would keep working nicely. Or consider cases where we do OriginalDefinition to then do things like navigating -- the nullable annotation doesn't matter there.

Although 100% not sold on that, it's just a hunch right now. I haven't actually moved any code that I know of that's using OriginalDefinition so I could totally imagine switching the moment I see an actual scenario.

Reply:
I'm not certain either.  I think this should potentially be something that comes up in a discussion with the relevant stakeholders (probably Gafter/Couv/Fred/You).  Basically: if we were to represent this in the ITypeSymbol itself, what do we think the lattice of all these pointers would look like.

My intuition is to look at something like `int?`.  Orig-def for that is `T?` not `T`.  I like *mostly* thinking about things the same way for nullable-reference-types versus nullable-value-types, but i don't feel super strong about that.

Reply:
I think as much as they somewhat look the same the nullable reference types and nullable structs do really behave differently in that `int?` is really a fundamentally different type versus just an annotation. In the case of `string?`, OriginalDefinition can't return `T?` since the concept simply doesn't exist so no matter what things are going to be have differently. And `IEnumerable<int?>` and `IEnumerable<int>` are fundamentally different types whereas the reference type annotation is just an annotation and that's it. It's...funky.

But yes very much looking forward to seeing what the first real use of OriginalDefinition needs to use that to guide.

Reply:
totall agree.  I look forward to seeing what we learn here :)